datastore/api/snippets.py

```diff
@@ -0,0 +1,820 @@
+    # [END multi_sort]
+
+    return tasks
+
+
+def keys_only_query(client):
+    # Create the entity that we're going to query.
+    upsert(client)
+
+    # [START keys_only_query]
+    query = client.query()
+    query.keys_only()
+    keys = list([x.key for x in query.fetch()])
+    # [END keys_only_query]
+
+    return keys
+
+
+def distinct_query(client):
+    # Create the entity that we're going to query.
+    upsert(client)
+
+    # [START distinct_query]
+    query = client.query(kind='Task')
+    query.group_by = ['type', 'priority']
+    query.order = ['type', 'priority']
+    tasks = list(query.fetch())
+    # [END distinct_query]
+
+    return tasks
+
+
+def distinct_on_query(client):
+    # Create the entity that we're going to query.
+    upsert(client)
+
+    # [START distinct_on_query]
+    query = client.query(kind='Task')
+    query.group_by = ['type']
+    query.order = ['type', 'priority']
+    tasks = list(query.fetch())
+    # [END distinct_on_query]
+
+    return tasks
+
+
+def kindless_query(client):
+    # [START kindless_query]
+    an_hour_ago = str(time.time() - 3600)
+    last_seen_key = client.key('Task', an_hour_ago)
+    query = client.query()
+    query.add_filter('__key__', '>', last_seen_key)
+    entities = list(query.fetch())
+    # [END kindless_query]
+
+    return entities
+
+
+def inequality_range(client):
+    # [START inequality_range]
+    query = client.query(kind='Task')
+    query.add_filter(
+        'created', '>', datetime.datetime(1990, 1, 1))
+    query.add_filter(
+        'created', '<', datetime.datetime(2000, 12, 31, 23, 59, 59))
+    tasks = list(query.fetch())
+    # [END inequality_range]
+
+    return tasks
+
+
+def inequality_invalid(client):
+    try:
+        # [START inequality_invalid]
+        query = client.query(kind='Task')
+        query.add_filter(
+            'created', '>', datetime.datetime(1990, 1, 1))
+        query.add_filter(
+            'priority', '>', 3)
+        tasks = list(query.fetch())
+        # [END inequality_invalid]
+
+        return tasks
+
+    except gcloud.exceptions.BadRequest:
+        pass
+
+
+def equal_and_inequality_range(client):
+    # [START equal_and_inequality_range]
+    query = client.query(kind='Task')
+    query.add_filter('priority', '=', 4)
+    query.add_filter('done', '=', False)
+    query.add_filter(
+        'created', '>', datetime.datetime(1990, 1, 1))
+    query.add_filter(
+        'created', '<', datetime.datetime(2000, 12, 31, 23, 59, 59))
+    tasks = list(query.fetch())
+    # [END equal_and_inequality_range]
+
+    return tasks
+
+
+def inequality_sort(client):
+    # [START inequality_sort]
+    query = client.query(kind='Task')
+    query.add_filter('priority', '>', 3)
+    query.order = ['priority', 'created']
+    tasks = list(query.fetch())
+    # [END inequality_sort]
+
+    return tasks
+
+
+def inequality_sort_invalid_not_same(client):
+    try:
+        # [START inequality_sort_invalid_not_same]
```

Comment:
To show that you can't do this kind of query (you can't use an inequality filter on one property and sort on a different one)